### ISSUE TITLE
Pipeline API: When using the Java language, the input of the AbstractInferenceModel.predict function is incorrect, resulting in a model prediction error loaded by the loadTF function.

### ISSUE THREAD

**Issue by LB-Yu** [intel-analytics/analytics-zoo#1615]
Hi!
I am using the Java language to call zoo's **pipeline API**. My goal is to use it to load the model and predict an image that is being input. I successfully loaded the model in the Tensorflow **SavedModel** format using the *loadTF* function. 

However, when I use the loaded model for image classification prediction, I always get the wrong result. The code I read the picture is as follows (I have pre-processed the image into a 256*256 size). The length of *fpixels* is 196608 (256 * 256 * 3), its order is the same as the order after the numpy three-dimensional array reshape (-1).
```java
public class ImageProcess {
      public static List<List<JTensor> > readExampleImage() throws IOException {
        String imagePath = System.getenv("IMAGE_PATH");
        BufferedImage image = ImageIO.read(new File(imagePath));
        Raster raster = image.getData();

        float[] ftmp = new float[raster.getWidth() * raster.getHeight() * raster.getNumBands()];
        float[] fpixels = raster.getPixels(0, 0, raster.getWidth(), raster.getHeight(), ftmp);

        JTensor tensor = new JTensor(fpixels, new int[]{1, 256, 256, 3});
        List<JTensor> batch = new ArrayList<>();
        batch.add(tensor);

        List<List<JTensor> > ret = new ArrayList<>();
        ret.add(batch);
        return ret;
    }
}
```
The code I loaded the model and predicted is as follows.
```java
public class ZooExample {

    public static class ClassificationModel extends AbstractInferenceModel {
        public ClassificationModel() {
            super();
        }
    }

    public static void main(String[] args) throws IOException {
        String modelPath = System.getenv("CLASSIFICATION_MODEL_PATH");
        boolean ifReverseInputChannels = true;
        int[] inputShape = {1, 256, 256, 3};
        float[] meanValues = {0.f, 0.f, 0.f};
        float scale = 1.0f;
        String input = "input";

        ClassificationModel model = new ClassificationModel();
        model.loadTF(modelPath, inputShape, ifReverseInputChannels, meanValues, scale, input);
        List<List<JTensor> > inputImage = ImageProcess.readExampleImage();
        List<List<JTensor>> result = model.predict(inputImage);
    }
}
```
The returned *result* contains only one **JTensor** of length 50 (I have 50 categories). But the content of **JTensor** is completely different from that predicted by Tensorflow (Tensorfow predicts the result is correct).

I want to know if I have misorganized the input of the predict function. How to organize *List<List<JTensor>>* when I need to predict a picture.

I would be very grateful if I could get help from anyone.

**Comment by qiyuangong** [comment]
Hi @LB-Yu 
I think your basic workflow of using `loadTF` and predicting with JTensor is correct. But, `meanValues` and `scale` of  `loadTF` maybe not set correctly. 
```java
float[] meanValues = {0.f, 0.f, 0.f};
float scale = 1.0f;
```

These two values are related to `normalization` in image pre-processing. You can find more examples about setting these two parameters from [Converting a TensorFlow* Model](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_Convert_Model_From_TensorFlow.html).

**Comment by LB-Yu** [comment]
Thanks for your answer!
But after I set the ***meanValues*** and ***scale*** correctly, the prediction result is still incorrect.  So I think the point of the problem doesn't seem to be here. The point I want to confirm is whether the order of ***JTensor*** is the same as the order of ***numpy.ndarray.reshape(-1)***.

In addition, I have another question, ***JTensor*** can represent a batch number of pictures, then ***List<JTensor>*** can represent a batch list, so why the input of the predict function should be set to ***List<List<JTensor>>***, I think ***JTensor*** already enough.

**Comment by qiyuangong** [comment]
Hi @LB-Yu 
We used `array.flatten()` which is equal to `reshape(-1)` in default order.  Can you provide more details about correct result and incorrect result?

Back to meanValues and scale. Actually, that values are related to `inputShape` and `ifReverseInputChannels `. It's totally different from TensorFlow. Here is an example about [Resnet_50](https://github.com/intel-analytics/analytics-zoo/blob/master/zoo/src/main/scala/com/intel/analytics/zoo/examples/vnni/openvino/PrepareOpenVINOResNet.scala#L62). Pls double check it.

@glorysdj for `List<List>` problem of `JTensor` .

**Comment by glorysdj** [comment]
Hi @LB-Yu, could you please paste the output.
The reason for List<List<JTensor>> is: the first List is for batch, e.g 5 different inputs; the second List is for table, you can refer to bigdl Table, e.g. sometimes there are two output JTensor. In your case, the batch and output table are both bunber 1, so you can get result with result.get(0).get(0)

**Comment by LB-Yu** [comment]
Hello!
I trained a tensorflow model and save it as *'saved model'* format. Then I load it with tensorflow as follows.
```python
from tensorflow.contrib import predictor
import numpy as np
from PIL import Image
import pprint

model_path = "model_path"

img = Image.open("example.jpg")
img = img.resize((224, 224))
img_arr = np.asarray(img)
pprint.pprint(img_arr.reshape(-1)[:100])

predict_fn = predictor.from_saved_model(model_path)
predictions = predict_fn({"input": np.reshape(img_arr, [1, 224, 224, 3])})

res = predictions['fc100'][0]
print(res)
label = np.argmax(res)
print(label)
```
I can get the following output. The result 51 obtained is correct.
```
[ 7 44  1 12 47  7 15 45  9 12 36  4  5 21  0  4 13  0 12 14  0 21 19  7
 15 11  2 15 11  2 20 17  8 25 27 14 23 30 14 15 24  5  5 18  0  1 16  0
 12 20  0 16 26  2 28 44 17  0 24  0  1 33  0 32 67 25 22 58 14 22 55 10
 36 63 20 43 66 22 44 65 22 25 46  3  0 13  0 27 54 11 28 59 17 24 57 14
 48 76 25 27]
[1.7684283e-04 4.4735093e-03 1.7239581e-03 1.1964597e-04 3.5377250e-06
 1.7435659e-05 5.3734431e-04 7.2625859e-05 1.0751063e-05 4.7659283e-05
 4.0124029e-05 3.3639776e-06 2.0891648e-05 8.6950626e-05 3.0142700e-03
 1.5665486e-05 1.9155198e-06 1.6262599e-03 7.0686487e-04 1.9399430e-04
 1.2166968e-05 2.1585161e-03 7.6889104e-05 8.7506156e-03 1.6028070e-03
 1.4431654e-03 2.4011662e-02 4.3043192e-03 3.8808089e-06 2.6393136e-06
 1.7025187e-03 1.7058317e-06 1.4115096e-04 5.4181059e-05 2.2182118e-02
 1.3596586e-04 2.4982481e-05 7.4555929e-04 5.7246390e-04 1.8190818e-03
 3.4037710e-04 2.0841879e-03 5.0427221e-05 1.6323211e-05 1.1398829e-05
 5.6353572e-05 4.2807581e-04 1.5990830e-05 3.7684054e-05 5.2446169e-03
 1.6742161e-03 4.3839979e-01 9.4868667e-02 1.9572253e-04 4.2584364e-04
 7.6807897e-05 1.5669230e-04 1.4709219e-07 5.2260731e-07 1.5694115e-04
 1.5642585e-05 2.2497133e-03 3.5716829e-04 2.7314074e-02 3.7001606e-04
 1.0302351e-04 1.9186208e-02 3.1912137e-02 9.6189418e-05 2.7610844e-03
 7.1375052e-06 4.1716921e-04 8.2646824e-05 1.9817057e-04 3.9836166e-05
 2.5966449e-07 6.3468760e-05 7.1383023e-04 1.1612709e-04 7.1869203e-05
 7.7113807e-05 9.8445125e-05 4.3123108e-03 1.0519107e-01 5.0738390e-04
 1.4775629e-04 1.9199038e-02 4.6002891e-04 4.0908853e-05 7.2502131e-05
 6.0770281e-02 4.6906370e-04 3.6418083e-04 5.4373795e-06 8.2842566e-02
 1.1061250e-03 4.5830691e-03 8.4105659e-06 7.4295116e-05 6.7135477e-03]
51
```
When I use Java load the model as follows.
```Java
public class ZooExample {

    public static class GarbageClassificationModel extends AbstractInferenceModel {
        public GarbageClassificationModel() {
            super();
        }
    }

    public static void main(String[] args) throws IOException {
        String modelPath = System.getenv("IMAGE_MODEL_PATH");
        boolean ifReverseInputChannels = true;
        int[] inputShape = {1, 224, 224, 3};
        float[] meanValues = {123.68f, 116.78f, 103.94f};
        float scale = 1.0f;
        String input = "input";

        GarbageClassificationModel model = new GarbageClassificationModel();
        model.loadTF(modelPath, inputShape, ifReverseInputChannels, meanValues, scale, input);

        List<List<JTensor> > inputImage = ImageProcess.readExampleImage();
        List<List<JTensor>> result = model.predict(inputImage);

        float[] oneHot = result.get(0).get(0).getData();
        System.out.println("one hot length: " + oneHot.length);
        Utils.printArray(oneHot, oneHot.length);
        float max = oneHot[0];
        int label = 0;
        for (int i = 1; i < 100; i++) {
            if (oneHot[i] > max) {
                max = oneHot[i];
                label = i;
            }
        }
        System.out.println("max: " + max);
        System.out.println("predict result: " + label);
    }
}
```
I got the following result. The result of Softmax is wrong.
```
Softmax Result:
5.920905E-5 2.0302874E-4 6.7337806E-4 1.4917413E-4 2.8743155E-4 0.0011672428 4.1750746E-4 0.0013993934 0.0014839522 0.7453513 2.9934364E-4 2.3591465E-6 4.2256732E-5 1.19898395E-4 0.0017931969 4.5181718E-5 1.5594813E-7 1.1746862E-4 1.7089072E-4 0.0022047465 0.0060128905 3.8149735E-4 0.001147427 3.5845364E-6 2.1488071E-5 0.03459902 8.306652E-6 1.1860057E-4 4.2756493E-4 2.3387326E-4 0.001222208 2.4870914E-7 5.498001E-6 0.001378397 1.2907409E-4 1.1601058E-5 1.5778385E-5 8.842406E-5 1.19832315E-4 1.1475347E-6 6.3791376E-4 0.0025334714 3.4464263E-6 2.5943265E-4 4.2202382E-4 2.8021745E-7 6.839975E-5 0.0033473703 7.681448E-5 1.6247582E-4 9.556326E-5 0.025850726 0.015623315 1.2322903E-6 0.0015785008 8.0319645E-4 3.2896694E-6 1.637359E-6 5.5774226E-8 5.863726E-4 7.621532E-7 2.3122175E-4 9.688653E-5 0.0140210595 7.929941E-5 2.939345E-6 0.0061282455 0.0018806239 1.0801084E-6 3.9174396E-4 1.5059172E-5 3.991488E-5 0.016243368 2.670309E-5 1.5367095E-4 2.2412048E-9 7.7165E-5 0.046528425 1.1128496E-4 0.0051182904 5.337349E-6 3.3912296E-5 1.5023662E-4 1.7834168E-4 2.8717166E-5 3.3999098E-4 0.012164155 6.734551E-4 3.550509E-5 8.611745E-5 0.019124094 7.399756E-5 7.726677E-5 7.551199E-6 0.0119407615 4.6286354E-5 1.6033043E-4 8.642351E-5 6.5820816E-5 0.009604852 max: 0.7453513

predict result:
9
```

**Comment by glorysdj** [comment]
@LB-Yu Would you please also check the input data, please make sure the pre-processed results are exactly the same. Thanks.

**Comment by qiyuangong** [comment]
@LB-Yu 

Pls check image layout, i.e., NCHW or NHWC. In your current input shape, i.e., `1, 224, 224, 3`. Input data should be in NHWC. If you use `1, 3, 224, 224`, then it should be NCHW.

**Comment by LB-Yu** [comment]
@qiyuangong 
I can confirm that the input to the model is absolutely N*224*224*3. If I change the input shape to N*3*224*224 I will got the following error.
```shell
Exception in thread "main" com.intel.analytics.zoo.pipeline.inference.InferenceRuntimeException: Openvino optimize tf image classification model error: 1, [ ERROR ]  MatMul wasn't able to infer shape because input dimensions are not compatible,[ ERROR ]  Shape is not defined for output 0 of "fc100/MatMul".,[ ERROR ]  Cannot infer shapes or values for node "fc100/MatMul".,[ ERROR ]  Not all output shapes were inferred or fully defined for node "fc100/MatMul". ,[ ERROR ]  ,[ ERROR ]  It can happen due to bug in custom shape infer function <function tf_matmul_infer at 0x7f39fc82a268>.,[ ERROR ]  Or because the node inputs have incorrect values/shapes.,[ ERROR ]  Or because input shapes are incorrect (embedded to the model or passed via --input_shape).,[ ERROR ]  Run Model Optimizer with --log_level=DEBUG for more information.,[ ERROR ]  Stopped shape/value propagation at "fc100/MatMul" node.
```
And I tried the RGB and BGR channel order, and the prediction results are not correct. I have confirmed that the input I thought was the correct order of the array.

**Comment by qiyuangong** [comment]
@LB-Yu 

Pls try this configuration:
```
int[] inputShape = {1, 224, 224, 3};
```
Then, covert image into NCHW. Here is an example function:
```scala
def fromHWC2CHW(data: Array[Float]): Array[Float] = {
    val resArray = new Array[Float](3 * 224 * 224)
    for (h <- 0 to 223) {
      for (w <- 0 to 223) {
        for (c <- 0 to 2) {
          resArray(c * 224 * 224 + h * 224 + w) = data(h * 224 * 3 + w * 3 + c)
        }
      }
    }
    resArray
  }
```

**Comment by LB-Yu** [comment]
Thank you so much! I finally solve my problem.
